### ISSUE TITLE
[Bug][KV Offload] --kv-offloading-size allocates 1.92x the requested host memory as unreclaimable shmem

### ISSUE THREAD

**Issue by yifjiang** [vllm-project/vllm#53042]
### Summary

`--kv-offloading-size N` allocates **1.92 × N** of host memory, and the excess is unreclaimable `Shmem`. On a memory-limited deployment this both wastes ~half the pool budget and turns a nominally-fitting configuration into an OOMKill.

Measured on a 4-node GB200 deployment (DP=4, TP=4 per node, hybrid GDN+GQA model with 4 KV cache groups, `--kv-cache-dtype fp8`, `--mamba-ssm-cache-dtype bfloat16`, `--block-size 16`), vLLM 0.26.0.

### Measurement — the factor is exactly linear

| `--kv-offloading-size` | per-rank mapping | node `Shmem` | ratio |
|---|---|---|---|
| 400 GiB | 192.2 GiB | 769.0 GiB | **1.922×** |
| 200 GiB |  96.2 GiB | 385.0 GiB | **1.925×** |

Halving the pool halves the footprint exactly, so this is a constant multiplicative factor rather than a fixed overhead or a size-dependent effect.

At `--kv-offloading-size 200`, from the worker cgroup at idle:

```
memory.current   443.5 GiB
shmem            385.0 GiB     <- the pool
anon              35.0 GiB     <- python, torch, CUDA ctx, NCCL
file - shmem       ~3 GiB      <- actual page cache
SwapTotal              0
```

The pool is ~87% of the pod's memory, and with no swap the `Shmem` can never be reclaimed — the cgroup OOMKills rather than evicting.

### Allocation shape

Each of the 4 TP ranks maps **24 separate 4.0 GiB `/dev/zero` `MAP_SHARED` regions, each with a distinct inode** — 24 × 4 GiB = 96 GiB per rank, × 4 ranks = 384 GiB. `smaps_rollup` shows `Pss ≈ Rss` and `Shared_Dirty = 0` on every rank, so none of these regions is shared between ranks.

The correct per-rank share of a 200 GiB pool with `world_size = 4` is 50 GiB. Each rank is holding **1.92× its share**.

This appears to contradict the design in `vllm/v1/kv_offload/cpu/shared_offload_region.py`, where `_worker_offset = rank * cpu_page_size` implies one shared region per node carved into per-rank slices, not N private regions per rank.

### What I ruled out

- **Not `world_size` being wrong.** `ParallelConfig(tensor_parallel_size=4, pipeline_parallel_size=1, data_parallel_size=4, data_parallel_external_lb=True).world_size` returns `4` for this deployment (checked all three variants).
- **Not page cache.** Actual page cache is ~3 GiB; the 385 GiB is `Shmem`.
- **Not size-dependent.** The two-point measurement above is linear.

I have not isolated the responsible line, so I am reporting the measurement rather than proposing a fix. Note `2 × (1 − 0.0435) = 1.913`, and the engine logs `Add 3 padding layers, may waste at most 4.35% KV cache memory` for this model — a 2× over-allocation net of that padding would land almost exactly on the observed 1.92×, but I have not confirmed the connection.

### Impact

On a pod with an 800 GiB limit:

| nominal | actual | fits? |
|---|---|---|
| 384 GiB | 737 GiB | yes, ~23 GiB headroom |
| 512 GiB | 983 GiB | **no — exceeds the limit outright** |

`--kv-offloading-size 512` OOMKilled the deployment during CUDA-graph capture. The pool alone was larger than the pod limit, so it could never have started; the practical ceiling is ~385 GiB nominal. Correcting the factor would roughly double the usable host tier for the same memory budget.

### Repro

Run any TP>1 deployment with `--kv-offloading-size N` and compare `N` against `grep shmem /sys/fs/cgroup/memory.stat`, or sum the `/dev/zero` mappings in `/proc/<worker-pid>/maps`.

**Comment by yifjiang** [comment]
## Root cause found: per-tensor allocations land 4% over a 2 GiB boundary and the pinned-host caching allocator rounds them to 4 GiB

vLLM's sizing math is **correct**. The waste is entirely allocator rounding.

### What the allocator actually does

With `VLLM_LOGGING_LEVEL=DEBUG`, `gpu_worker.py:503` reports the real allocations. Per rank, on a 200 GiB pool with `world_size=4`:

```
24 tensors x  torch.zeros((2053, 1089536), pin_memory=True)
            = 2,236,817,408 B = 2.0832 GiB each
24 x 2.0832 GiB = 50.0 GiB  ==  cpu_bytes_to_use / world_size    <- correct
```

But each of those maps to a **4.0 GiB** region (24 x 4 GiB = 96 GiB/rank, x4 ranks = 384 GiB).

### Standalone repro

```python
import torch
def shmem():
    for l in open("/proc/meminfo"):
        if l.startswith("Shmem:"): return int(l.split()[1])/1048576

torch.zeros(1, device="cuda")          # init ctx first
PAGE = 1089536
for name, nblk in [("2.0832 GiB (as vLLM allocates)", 2053),
                   ("1.9996 GiB (just under 2 GiB)", 1971)]:
    b = shmem()
    t = torch.zeros((nblk, PAGE), dtype=torch.int8, device="cpu", pin_memory=True)
    print(name, "->", round(shmem() - b, 3), "GiB")
    del t
```

Result on GB200 / torch 2.11:

| requested | Shmem delta | ratio |
|---|---|---|
| **2.0832 GiB** (vLLM's size) | **4.006 GiB** | **1.923x** |
| 1.9996 GiB (4% smaller) | 2.000 GiB | 1.000x |
| 1.0 GiB | 1.000 GiB | 1.001x |

`1.923x` in this 12-line repro matches the `1.92x` measured on the live 4-node deployment to three digits. **Shrinking the request by 4% removes 100% of the waste.**

### Why it is never the exact-size path

`SharedOffloadRegion` (which `ftruncate`s an exact size into `/dev/shm` and shares one region per node) is only ever constructed in `v1/kv_offload/tiering/spec.py`. `CPUOffloadingSpec` never passes an `mmap_region`, so it always takes the `torch.zeros(..., pin_memory=True)` fallback at `gpu_worker.py:497` and always pays the rounding. That also explains the observed shape: 24 *private* `/dev/zero` mappings per rank rather than one shared `/dev/shm/vllm_offload_*.mmap`.

### Suggested fix

Clamp `num_cpu_blocks` so each per-tensor allocation does not cross a power-of-two boundary — i.e. round `num_cpu_blocks * cpu_page_size_bytes` **down** to the largest power of two that fits. Costs at most the rounding loss in capacity (here 4%) and saves up to 2x host memory. Alternatively, route `CPUOffloadingSpec` through `SharedOffloadRegion` as the tiering spec already does, which sidesteps the allocator entirely and shares one region per node.

### Workaround for affected deployments

`kv_connector_extra_config: {"spec_name": "TieringOffloadingSpec"}` selects the spec that uses the exact-size shared region (`secondary_tiers` is optional, so it runs CPU-only). Untested by me so far, but it is the registered path that avoids the fallback.

Corrections to the original report, now that the mechanism is known: the measurements were taken on **0.26.0**, which differs from `main` in two relevant ways — `BLOCK_SIZE_ALIGNMENT = 1` (no padding) and no `replicated_layout`. The `2 x (1 - 0.0435)` padding coincidence I noted there is **not** the mechanism; the power-of-two rounding is. `main`'s `CPUOffloadingSpec` still has the same fallback, so I believe it is still affected, but the exact per-tensor size there may differ.

**Comment by yifjiang** [comment]
## Correction: the cost is `next_pow2(per-tensor)`, a sawtooth — not the linear 1.92x I reported

My original two measurements were at **400 GiB and 200 GiB — exactly one octave apart**, which is precisely the case where a power-of-two step function is indistinguishable from a constant factor. That was a bad experiment design on my part, and "1.92x linear" is wrong.

Sweeping per-tensor size on the same hardware (ignoring rows where the caching allocator reused an already-freed block, which report a 0 delta):

| per-tensor requested | Shmem delta | ratio |
|---|---|---|
| 1.3333 GiB | 2.000 GiB | 1.500x |
| 2.0832 GiB | 4.000 GiB | 1.920x |
| 4.1664 GiB | 8.000 GiB | 1.920x |

The allocator rounds each request **up to the next power of two**. So the waste depends entirely on where the per-tensor size falls, and is **zero** when it lands on or just under a boundary.

### What this means in practice

Per-tensor size is `cpu_bytes_to_use / (world_size * n_kv_tensors)`. For the deployment in the original report (`world_size=4`, 24 tensors, i.e. `nominal/96`):

| nominal | per-tensor | rounds to | cost |
|---|---|---|---|
| 192 GiB | 2.0000 GiB | 2 GiB | **1.00x** |
| 200 GiB | 2.0832 GiB | 4 GiB | 1.92x |
| 384 GiB | 4.0000 GiB | 4 GiB | **1.00x** |
| 400 GiB | 4.1664 GiB | 8 GiB | 1.92x |
| 512 GiB | 5.3333 GiB | 8 GiB | 1.50x |

The 512 GiB case is what OOMKilled our deployment: 512 x 1.50 = 768 GiB plus ~40 GiB anon exceeds an 800 GiB pod limit. The 1.50x there is consistent with this model and **not** with the 1.92x I originally claimed.

So the impact statement in the original report was wrong: a deployment sized at 384 GiB pays nothing, and the efficient sizes are quantised to `n_tensors * world_size * 2^k`. The bug is real but it is a **sawtooth penalty on unlucky sizes**, not a flat doubling.

### The workaround works

`kv_connector_extra_config: {"spec_name": "TieringOffloadingSpec"}` avoids it entirely, as predicted — that spec allocates through `SharedOffloadRegion` (exact `ftruncate`) instead of the rounding allocator. Measured on a single node, 8 GiB pool, same image, one variable:

| spec | Shmem delta | ratio | `torch.zeros pinned tensor` allocations |
|---|---|---|---|
| `CPUOffloadingSpec` (default) | 12.24 GiB | 1.530x | 12 |
| `TieringOffloadingSpec` | 8.24 GiB | **1.029x** | **0** |

This also removes the size quantisation, which matters more than the wasted bytes: with the default spec, usable pool sizes are effectively restricted to `n_tensors * world_size * 2^k` (for us: 96, 192, 384, 768 GiB — nothing usable between 384 and 768), whereas the tiering path accepts any size at ~1.03x.

Suggested fix stands, with the emphasis shifted: round `num_cpu_blocks` **down** so each per-tensor allocation lands at or under a power of two, or route `CPUOffloadingSpec` through `SharedOffloadRegion` as the tiering spec already does. The latter also fixes the quantisation.

**Comment by AbarnaaSree** [comment]
I’d like to investigate this issue, particularly the `CPUOffloadingSpec` allocation path and its interaction with the pinned-memory allocator. I can compare it with the existing `SharedOffloadRegion` approach and determine whether adjusting the allocation size or reusing the exact-size shared region would be the cleaner fix. If the maintainers are open to it, I’d be happy to work on a PR with regression coverage.
